pkg/utils/helper.go

```diff
@@ -0,0 +1,74 @@
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+	"gopkg.in/yaml.v2"
+)
+
+// function to get the relevant information out of k8s object that is needed to compute hash
+func GetHashObjectOfKubernetes(object interface{}) (map[string]interface{}, error) {
+	// By marshaling/unmarshaling the object via JSON we're copying it into a
+	// map, which is easier to manipulate in generic way than structs.
+	if b, err := json.Marshal(object); err != nil {
+		return nil, fmt.Errorf("Error while encoding %+v into JSON: %s", object, err)
+	} else {
+		var v map[string]interface{}
+		if err := json.Unmarshal(b, &v); err != nil {
+			return nil, fmt.Errorf("Error while decoding JSON %s into an object: %s", v, err)
+		} else {
+			// We need only kind/version/spec and labels excluding the label with the
+			// current hash value while calculating the hash.  Also Kubernetes adds
+			// its own info into `metadata` which we need to ignore.
+			meta := v["metadata"].(map[string]interface{})
+			labels := meta["labels"]
+			if labels != nil {
+				delete(labels.(map[string]interface{}), "yelp.com/operator_config_hash")
+			} else {
+				labels = make(map[string]interface{})
+			}
+			m := map[string]interface{}{
+				"kind":       v["kind"],
+				"apiVersion": v["apiVersion"],
+				"spec":       v["spec"],
+				"metadata": map[string]interface{}{
+					"name":      meta["name"],
+					"namespace": meta["namespace"],
+					"labels":    labels,
+				},
+			}
+			return m, nil
+		}
+	}
+}
+
+// function to get the yaml output of an object
+func GetYamlOfObject(object interface{}) (string, error) {
+	if b, err := json.Marshal(object); err != nil {
+		return "", fmt.Errorf("Error while encoding %+v into JSON: %s", object, err)
+	} else {
+		var v map[string]interface{}
+		if err := json.Unmarshal(b, &v); err != nil {
+			return "", fmt.Errorf("Error while decoding JSON %s into an object: %s", v, err)
+		} else {
+			if y, err := yaml.Marshal(&v); err != nil {
+				return "", fmt.Errorf("Error while encoding Map %s into YAML: %s", v, err)
```

Comment:
Is there any reason why we need this particular function to go through cycle marshal/unmarshal/marshal ? We only use it to generate YAML from an object which is a result of this cycle already (from `GetHashObjectOfKubernets`) and I think that should be enough?

If we do *not* need this cycle here, then let's get rid of this function and let `GetYamlOfHashObjectOfK8sObject` use `yaml.Marshal` directly. If we *do need* them, then I'd like to see a comment here to explain why direct use of `yaml.Marshal` is not sufficient, in the context of `GetYamlOfHashObjectOfK8sObject`

Reply:
I added this function just to provide the functionality to get yaml of any go object. Anyone who uses the library and need to get the yaml for their object can directly use this function.

Reply:
But that is what `yaml.Marshal` already does, so we do not need this function. At least, not for objects which already went through the cycle of marshal/unmarshal.

Reply:
I did it this way as I wanted to remove the null/nil fields of the object. When I do marshal/unmarshal into json null/nil fields are removed. Let me check how to achieve the same with yaml.Marshal

Reply:
I checked since we already did marshal/unmarshal for the object in `GetHashObjectOfKubernets`  we don't need it for that. For cases where we directly call this method, the yaml we get from yaml.marshal vs when we first marshal/unmarshal the object to get yaml, the results are different. All the existing appsV1 structs have field tags only supported for json marshal/unmarshal. The same kind of result can be achieved with yaml if the appsV1 structs add support for yaml marshal/unmarshal.

Reply:
But we do not need to call this method anywhere right now. So if `yaml.Marshal` without preceding cycle marshal/unmarshal does not deliver the right results, that does not affect anyone.

Reply:
Ahhh I believe if we are adding a support to get yaml of go object, it should be similar to what one expect from json. This is what my thought process was, if you want I can remove it and change it accommodate what the current usage is.

Reply:
Yes, you got it exactly right - I want the code to accommodate the current usage, rather than hypothetical more generic case. Thank you!